File: gin/internal/model/content/article.go
package model_content

import (
	"time"

	"gorm.io/gorm"
	"fmt"
)

type ArticleStatusType string
const (
	ArticleStatusType_Draft ArticleStatusType = "Draft"
	ArticleStatusType_Published ArticleStatusType = "Published"
	ArticleStatusType_Archived ArticleStatusType = "Archived"
)

type ArticleArticleTypeType string
const (
	ArticleArticleTypeType_Guide ArticleArticleTypeType = "Guide"
	ArticleArticleTypeType_Tierlist ArticleArticleTypeType = "Tierlist"
	ArticleArticleTypeType_Matchup ArticleArticleTypeType = "Matchup"
	ArticleArticleTypeType_News ArticleArticleTypeType = "News"
	ArticleArticleTypeType_Spotlight ArticleArticleTypeType = "Spotlight"
	ArticleArticleTypeType_Decklist ArticleArticleTypeType = "Decklist"
)

type ArticleLanguageType string
const (
	ArticleLanguageType_EN ArticleLanguageType = "EN"
	ArticleLanguageType_DE ArticleLanguageType = "DE"
	ArticleLanguageType_FR ArticleLanguageType = "FR"
	ArticleLanguageType_IT ArticleLanguageType = "IT"
	ArticleLanguageType_ES ArticleLanguageType = "ES"
	ArticleLanguageType_JP ArticleLanguageType = "JP"
	ArticleLanguageType_PT ArticleLanguageType = "PT"
)

// ArticleCreateRequest is the POST body.
type ArticleCreateRequest struct {
	Title string `json:"title" binding:"required"`
	Slug string `json:"slug" binding:"required"`
	Body string `json:"body" binding:"required"`
	Excerpt *string `json:"excerpt"`
	CoverImageUrl *string `json:"cover_image_url"`
	Status ArticleStatusType `json:"status" binding:"required"`
	ArticleType ArticleArticleTypeType `json:"article_type" binding:"required"`
	Language ArticleLanguageType `json:"language" binding:"required"`
	ViewCount int `json:"view_count"`
	LikesCount int `json:"likes_count"`
	IsFeatured bool `json:"is_featured"`
	PublishedAt *string `json:"published_at"`
	AuthorID uint `json:"author_id"`
	FeaturedDeckID *uint `json:"featured_deck_id"`
}

// ArticleUpdateRequest is the PUT/PATCH body — all fields optional.
type ArticleUpdateRequest struct {
	Title *string `json:"title"`
	Slug *string `json:"slug"`
	Body *string `json:"body"`
	Excerpt *string `json:"excerpt"`
	CoverImageUrl *string `json:"cover_image_url"`
	Status *ArticleStatusType `json:"status"`
	ArticleType *ArticleArticleTypeType `json:"article_type"`
	Language *ArticleLanguageType `json:"language"`
	ViewCount *int `json:"view_count"`
	LikesCount *int `json:"likes_count"`
	IsFeatured *bool `json:"is_featured"`
	PublishedAt *string `json:"published_at"`
	AuthorID *uint `json:"author_id"`
	FeaturedDeckID *uint `json:"featured_deck_id"`
}

// ArticleResponse is the JSON representation returned by the API.
type ArticleResponse struct {
	ID        uint      `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Title string `json:"title"`
	Slug string `json:"slug"`
	Body string `json:"body"`
	Excerpt *string `json:"excerpt"`
	CoverImageUrl *string `json:"cover_image_url"`
	Status ArticleStatusType `json:"status"`
	ArticleType ArticleArticleTypeType `json:"article_type"`
	Language ArticleLanguageType `json:"language"`
	ViewCount int `json:"view_count"`
	LikesCount int `json:"likes_count"`
	IsFeatured bool `json:"is_featured"`
	PublishedAt *string `json:"publishedAt"`
	AuthorID uint `json:"author_id"`
	FeaturedDeckID *uint `json:"featured_deck_id"`
}

type Article struct {
	gorm.Model
	Title string `gorm:"column:title;not null"`
	Slug string `gorm:"column:slug;not null;uniqueIndex"`
	Body string `gorm:"column:body;type:text;not null"`
	Excerpt *string `gorm:"column:excerpt;type:text"`
	CoverImageUrl *string `gorm:"column:cover_image_url"`
	Status ArticleStatusType `gorm:"column:status;not null;default:'Draft'"`
	ArticleType ArticleArticleTypeType `gorm:"column:article_type;not null;default:'Guide'"`
	Language ArticleLanguageType `gorm:"column:language;not null;default:'EN'"`
	ViewCount int `gorm:"column:view_count;not null;default:0"`
	LikesCount int `gorm:"column:likes_count;not null;default:0"`
	IsFeatured bool `gorm:"column:is_featured;default:false"`
	PublishedAt *string `gorm:"column:published_at"`
	AuthorID uint `gorm:"column:author_id;constraint:OnDelete:RESTRICT"`
	FeaturedDeckID *uint `gorm:"column:featured_deck_id;constraint:OnDelete:SET NULL"`
	TagAssignments []ArticleTagAssignment `gorm:"foreignKey:ArticleID"`
	Comments []ArticleComment `gorm:"foreignKey:ArticleID"`
}

func (m *Article) ToResponse() ArticleResponse {
	return ArticleResponse{
		ID:        m.ID,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
		Title: m.Title,
		Slug: m.Slug,
		Body: m.Body,
		Excerpt: m.Excerpt,
		CoverImageUrl: m.CoverImageUrl,
		Status: m.Status,
		ArticleType: m.ArticleType,
		Language: m.Language,
		ViewCount: m.ViewCount,
		LikesCount: m.LikesCount,
		IsFeatured: m.IsFeatured,
		PublishedAt: m.PublishedAt,
		AuthorID: m.AuthorID,
		FeaturedDeckID: m.FeaturedDeckID,
	}
}

func (m *Article) ApplyUpdate(req ArticleUpdateRequest) {
	if req.Title != nil { m.Title = *req.Title }
	if req.Slug != nil { m.Slug = *req.Slug }
	if req.Body != nil { m.Body = *req.Body }
	if req.Excerpt != nil { m.Excerpt = req.Excerpt }
	if req.CoverImageUrl != nil { m.CoverImageUrl = req.CoverImageUrl }
	if req.Status != nil { m.Status = *req.Status }
	if req.ArticleType != nil { m.ArticleType = *req.ArticleType }
	if req.Language != nil { m.Language = *req.Language }
	if req.ViewCount != nil { m.ViewCount = *req.ViewCount }
	if req.LikesCount != nil { m.LikesCount = *req.LikesCount }
	if req.IsFeatured != nil { m.IsFeatured = *req.IsFeatured }
	if req.PublishedAt != nil { m.PublishedAt = req.PublishedAt }
	if req.AuthorID != nil { m.AuthorID = *req.AuthorID }
	if req.FeaturedDeckID != nil { m.FeaturedDeckID = req.FeaturedDeckID }
}

// ── Lifecycle state machine ──────────────────────────────────────
var ArticleAllowedTransitions = map[string]map[string]bool{
		"Draft": {"Published": true},
		"Published": {"Archived": true},
		"Archived": {"Draft": true},
}

func (m *Article) AssertTransition(to string) error {
	current := string(m.Status)
	allowed, ok := ArticleAllowedTransitions[current]
	if !ok || !allowed[to] {
		return fmt.Errorf("transition %s -> %s is not allowed", current, to)
	}
	return nil
}

// ── Business operations ──────────────────────────────────────────

func (m *Article) Publish()  error {
	return fmt.Errorf("Publish: not implemented")
}

func (m *Article) Archive()  error {
	return fmt.Errorf("Archive: not implemented")
}

func (m *Article) IncrementView()  error {
	return fmt.Errorf("IncrementView: not implemented")
}

func (m *Article) Like()  error {
	return fmt.Errorf("Like: not implemented")
}

func (m *Article) Unlike()  error {
	return fmt.Errorf("Unlike: not implemented")
}

func (m *Article) ReadingTimeMinutes()  (int, error) {
	return 0, fmt.Errorf("ReadingTimeMinutes: not implemented")
}

File: axum/src/models/content/article.rs
#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum ArticleStatus {
    Draft,
    Published,
    Archived,
}

impl ArticleStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            ArticleStatus::Draft => "Draft",
            ArticleStatus::Published => "Published",
            ArticleStatus::Archived => "Archived",
        }
    }
}

impl std::fmt::Display for ArticleStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for ArticleStatus {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Draft" => Ok(ArticleStatus::Draft),
            "Published" => Ok(ArticleStatus::Published),
            "Archived" => Ok(ArticleStatus::Archived),
            _ => Err(format!("unknown ArticleStatus: {}", s)),
        }
    }
}

impl From<String> for ArticleStatus {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid ArticleStatus: {}", s))
    }
}

#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum ArticleArticleType {
    Guide,
    Tierlist,
    Matchup,
    News,
    Spotlight,
    Decklist,
}

impl ArticleArticleType {
    pub fn as_str(&self) -> &'static str {
        match self {
            ArticleArticleType::Guide => "Guide",
            ArticleArticleType::Tierlist => "Tierlist",
            ArticleArticleType::Matchup => "Matchup",
            ArticleArticleType::News => "News",
            ArticleArticleType::Spotlight => "Spotlight",
            ArticleArticleType::Decklist => "Decklist",
        }
    }
}

impl std::fmt::Display for ArticleArticleType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for ArticleArticleType {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Guide" => Ok(ArticleArticleType::Guide),
            "Tierlist" => Ok(ArticleArticleType::Tierlist),
            "Matchup" => Ok(ArticleArticleType::Matchup),
            "News" => Ok(ArticleArticleType::News),
            "Spotlight" => Ok(ArticleArticleType::Spotlight),
            "Decklist" => Ok(ArticleArticleType::Decklist),
            _ => Err(format!("unknown ArticleArticleType: {}", s)),
        }
    }
}

impl From<String> for ArticleArticleType {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid ArticleArticleType: {}", s))
    }
}

#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum ArticleLanguage {
    EN,
    DE,
    FR,
    IT,
    ES,
    JP,
    PT,
}

impl ArticleLanguage {
    pub fn as_str(&self) -> &'static str {
        match self {
            ArticleLanguage::EN => "EN",
            ArticleLanguage::DE => "DE",
            ArticleLanguage::FR => "FR",
            ArticleLanguage::IT => "IT",
            ArticleLanguage::ES => "ES",
            ArticleLanguage::JP => "JP",
            ArticleLanguage::PT => "PT",
        }
    }
}

impl std::fmt::Display for ArticleLanguage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for ArticleLanguage {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "EN" => Ok(ArticleLanguage::EN),
            "DE" => Ok(ArticleLanguage::DE),
            "FR" => Ok(ArticleLanguage::FR),
            "IT" => Ok(ArticleLanguage::IT),
            "ES" => Ok(ArticleLanguage::ES),
            "JP" => Ok(ArticleLanguage::JP),
            "PT" => Ok(ArticleLanguage::PT),
            _ => Err(format!("unknown ArticleLanguage: {}", s)),
        }
    }
}

impl From<String> for ArticleLanguage {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid ArticleLanguage: {}", s))
    }
}

#[derive(Debug, Clone, sqlx::FromRow, serde::Serialize)]
pub struct Article {
    pub id: i64,
    pub created_at: String,
    pub updated_at: String,
    pub title: String,
    pub slug: String,
    pub body: String,
    pub excerpt: Option<String>,
    pub cover_image_url: Option<String>,
    pub status: ArticleStatus,
    pub article_type: ArticleArticleType,
    pub language: ArticleLanguage,
    pub view_count: i64,
    pub likes_count: i64,
    #[serde(serialize_with = "crate::serde_utils::serialize_i64_as_bool")]
    pub is_featured: i64,
    #[serde(rename = "publishedAt")]
    pub published_at: Option<String>,
    pub author_id: i64,
    pub featured_deck_id: Option<i64>,
}

#[derive(Debug, serde::Deserialize)]
pub struct ArticleCreateRequest {
    pub title: String,
    pub slug: String,
    pub body: String,
    pub excerpt: Option<String>,
    pub cover_image_url: Option<String>,
    pub status: ArticleStatus,
    pub article_type: ArticleArticleType,
    pub language: ArticleLanguage,
    pub view_count: i64,
    pub likes_count: i64,
    pub is_featured: bool,
    pub published_at: Option<String>,
    pub author_id: i64,
    pub featured_deck_id: Option<i64>,
}

#[derive(Debug, serde::Deserialize)]
pub struct ArticleUpdateRequest {
    pub title: Option<String>,
    pub slug: Option<String>,
    pub body: Option<String>,
    pub excerpt: Option<String>,
    pub cover_image_url: Option<String>,
    pub status: Option<ArticleStatus>,
    pub article_type: Option<ArticleArticleType>,
    pub language: Option<ArticleLanguage>,
    pub view_count: Option<i64>,
    pub likes_count: Option<i64>,
    pub is_featured: Option<bool>,
    pub published_at: Option<String>,
    pub author_id: Option<i64>,
    pub featured_deck_id: Option<i64>,
}

impl Article {
    pub fn assert_transition(&self, to: &str) -> Result<(), String> {
        let allowed: &[(&str, &str)] = &[
            ("Draft", "Published"),
            ("Published", "Archived"),
            ("Archived", "Draft"),
        ];
        let current = &self.status;
        if allowed.iter().any(|(from, t)| *from == current.as_str() && *t == to) {
            Ok(())
        } else {
            Err(format!("transition {} -> {} not allowed", current, to))
        }
    }
}
